File: src/cdm_wizard/exporter.py
from openpyxl import Workbook
from openpyxl.styles import Font, PatternFill, Alignment, Border, Side
from openpyxl.utils import get_column_letter

from cdm_wizard.model import (
    FUNCTION_COLORS,
    GOVERN_FUNCTIONS,
    LEFT_OF_BOOM_FUNCTIONS,
    RIGHT_OF_BOOM_FUNCTIONS,
    asset_classes,
    functions,
)


def export_to_excel(data, output_path="cdm_output.xlsx"):
    """
    Exports CDM data to a styled Excel file.
    """
    function_list = functions()
    assets = [asset for asset in asset_classes() if asset in data]

    # Create workbook and sheet
    wb = Workbook()
    ws = wb.active
    ws.title = 'Cyber Defense Matrix'
    
    # Styles
    header_font = Font(bold=True, size=12)
    boom_font = Font(bold=True, size=14, color="FFFFFF")
    center_alignment = Alignment(horizontal='center', vertical='center', wrap_text=True)
    left_alignment = Alignment(horizontal='left', vertical='top', wrap_text=True)
    
    thin_border = Border(
        left=Side(style='thin'), 
        right=Side(style='thin'), 
        top=Side(style='thin'), 
        bottom=Side(style='thin')
    )

    # 1. Add Boom Headers
    govern_col = 2
    left_boom_start = govern_col + len(GOVERN_FUNCTIONS)
    left_boom_end = left_boom_start + len(LEFT_OF_BOOM_FUNCTIONS) - 1
    right_boom_start = left_boom_end + 1
    right_boom_end = right_boom_start + len(RIGHT_OF_BOOM_FUNCTIONS) - 1

    govern_cell = ws.cell(row=1, column=govern_col, value="GOVERN (Cross-Cutting)")
    govern_cell.fill = PatternFill(start_color="6AA84F", end_color="6AA84F", fill_type="solid")
    govern_cell.font = boom_font
    govern_cell.alignment = center_alignment
    govern_cell.border = thin_border

    ws.merge_cells(start_row=1, start_column=left_boom_start, end_row=1, end_column=left_boom_end)
    ws.merge_cells(start_row=1, start_column=right_boom_start, end_row=1, end_column=right_boom_end)

    left_boom_cell = ws.cell(row=1, column=left_boom_start, value="LEFT OF BOOM (Pre-Event)")
    left_boom_cell.fill = PatternFill(start_color="4F81BD", end_color="4F81BD", fill_type="solid")
    left_boom_cell.font = boom_font
    left_boom_cell.alignment = center_alignment

    right_boom_cell = ws.cell(row=1, column=right_boom_start, value="RIGHT OF BOOM (Post-Event)")
    right_boom_cell.fill = PatternFill(start_color="C0504D", end_color="C0504D", fill_type="solid")
    right_boom_cell.font = boom_font
    right_boom_cell.alignment = center_alignment

    # 2. Add Function Headers
    for col_idx, func in enumerate(function_list, start=2):
        cell = ws.cell(row=2, column=col_idx, value=func)
        cell.font = header_font
        cell.alignment = center_alignment
        cell.fill = PatternFill(start_color=FUNCTION_COLORS[func], end_color=FUNCTION_COLORS[func], fill_type="solid")
        cell.border = thin_border

    # 3. Add Asset Headers and Data
    for row_idx, asset in enumerate(assets, start=3):
        # Asset Name
        asset_cell = ws.cell(row=row_idx, column=1, value=asset)
        asset_cell.font = header_font
        asset_cell.alignment = center_alignment
        asset_cell.border = thin_border
        
        for col_idx, func in enumerate(function_list, start=2):
            cell_data = data[asset].get(func, {"Tech": "", "People": "", "Process": ""})
            value = f"TECH: {cell_data['Tech']}\nPEOPLE: {cell_data['People']}\nPROCESS: {cell_data['Process']}"
            
            cell = ws.cell(row=row_idx, column=col_idx, value=value)
            cell.alignment = left_alignment
            cell.border = thin_border
            cell.fill = PatternFill(start_color=FUNCTION_COLORS[func], end_color=FUNCTION_COLORS[func], fill_type="solid")

    # 4. Add Boom Line (Thick border between Protect and Detect)
    boom_boundary_col = 2 + function_list.index("Protect")
    for r in range(1, len(assets) + 3):
        existing_border = ws.cell(row=r, column=boom_boundary_col).border
        ws.cell(row=r, column=boom_boundary_col).border = Border(
            left=existing_border.left,
            right=Side(style='thick'),
            top=existing_border.top,
            bottom=existing_border.bottom
        )

    # 5. Add Dependency Legend at bottom
    legend_row = len(assets) + 4
    ws.merge_cells(start_row=legend_row, start_column=1, end_row=legend_row, end_column=len(function_list) + 1)
    legend_cell = ws.cell(
        row=legend_row,
        column=1,
        value="Dependency emphasis: Govern is process-led | Identify/Protect lean Technology | Detect/Respond/Recover lean People",
    )
    legend_cell.font = Font(italic=True)
    legend_cell.alignment = center_alignment

    # Auto-adjust column widths
    ws.column_dimensions['A'].width = 15
    for col_idx in range(2, len(function_list) + 2):
        ws.column_dimensions[get_column_letter(col_idx)].width = 28
        
    wb.save(output_path)
    return output_path

File: gosrc/internal/exporter/exporter.go
package exporter

import (
	"fmt"

	"cdmbuddy/internal/model"
	"github.com/xuri/excelize/v2"
)

func ExportToExcel(matrix model.Matrix, path string) error {
	f := excelize.NewFile()
	defer func() {
		if err := f.Close(); err != nil {
			fmt.Println(err)
		}
	}()

	sheetName := "Cyber Defense Matrix"
	f.SetSheetName("Sheet1", sheetName)

	// Styles
	headerStyle, _ := f.NewStyle(&excelize.Style{
		Font:      &excelize.Font{Bold: true, Size: 12},
		Alignment: &excelize.Alignment{Horizontal: "center", Vertical: "center", WrapText: true},
		Border: []excelize.Border{
			{Type: "left", Color: "000000", Style: 1},
			{Type: "top", Color: "000000", Style: 1},
			{Type: "bottom", Color: "000000", Style: 1},
			{Type: "right", Color: "000000", Style: 1},
		},
	})

	boomStyleBase := &excelize.Style{
		Font:      &excelize.Font{Bold: true, Size: 14, Color: "FFFFFF"},
		Alignment: &excelize.Alignment{Horizontal: "center", Vertical: "center", WrapText: true},
		Border: []excelize.Border{
			{Type: "left", Color: "000000", Style: 1},
			{Type: "top", Color: "000000", Style: 1},
			{Type: "bottom", Color: "000000", Style: 1},
			{Type: "right", Color: "000000", Style: 1},
		},
	}

	governBoomStyle, _ := f.NewStyle(&excelize.Style{
		Fill:      excelize.Fill{Type: "pattern", Color: []string{"6AA84F"}, Pattern: 1},
		Font:      boomStyleBase.Font,
		Alignment: boomStyleBase.Alignment,
		Border:    boomStyleBase.Border,
	})

	leftBoomStyle, _ := f.NewStyle(&excelize.Style{
		Fill:      excelize.Fill{Type: "pattern", Color: []string{"4F81BD"}, Pattern: 1},
		Font:      boomStyleBase.Font,
		Alignment: boomStyleBase.Alignment,
		Border:    boomStyleBase.Border,
	})

	rightBoomStyle, _ := f.NewStyle(&excelize.Style{
		Fill:      excelize.Fill{Type: "pattern", Color: []string{"C0504D"}, Pattern: 1},
		Font:      boomStyleBase.Font,
		Alignment: boomStyleBase.Alignment,
		Border:    boomStyleBase.Border,
	})

	// 1. Add Boom Headers
	f.SetCellValue(sheetName, "B1", "GOVERN (Cross-Cutting)")
	f.SetCellStyle(sheetName, "B1", "B1", governBoomStyle)

	f.MergeCell(sheetName, "C1", "D1")
	f.SetCellValue(sheetName, "C1", "LEFT OF BOOM (Pre-Event)")
	f.SetCellStyle(sheetName, "C1", "D1", leftBoomStyle)

	f.MergeCell(sheetName, "E1", "G1")
	f.SetCellValue(sheetName, "E1", "RIGHT OF BOOM (Post-Event)")
	f.SetCellStyle(sheetName, "E1", "G1", rightBoomStyle)

	// 2. Add Function Headers
	for i, funcName := range model.Functions {
		col, _ := excelize.ColumnNumberToName(i + 2)
		cell := col + "2"
		f.SetCellValue(sheetName, cell, funcName)
		
		color := model.FunctionColors[funcName]
		style, _ := f.NewStyle(&excelize.Style{
			Font:      &excelize.Font{Bold: true, Size: 12},
			Alignment: &excelize.Alignment{Horizontal: "center", Vertical: "center", WrapText: true},
			Fill:      excelize.Fill{Type: "pattern", Color: []string{color}, Pattern: 1},
			Border: []excelize.Border{
				{Type: "left", Color: "000000", Style: 1},
				{Type: "top", Color: "000000", Style: 1},
				{Type: "bottom", Color: "000000", Style: 1},
				{Type: "right", Color: "000000", Style: 1},
			},
		})
		f.SetCellStyle(sheetName, cell, cell, style)
	}

	// 3. Add Asset Headers and Data
	for r, asset := range model.AssetClasses {
		rowIdx := r + 3
		cellName := fmt.Sprintf("A%d", rowIdx)
		f.SetCellValue(sheetName, cellName, asset)
		f.SetCellStyle(sheetName, cellName, cellName, headerStyle)

		for c, funcName := range model.Functions {
			col, _ := excelize.ColumnNumberToName(c + 2)
			cell := fmt.Sprintf("%s%d", col, rowIdx)
			
			cellData := matrix[asset][funcName]
			value := fmt.Sprintf("TECH: %s\nPEOPLE: %s\nPROCESS: %s", cellData.Tech, cellData.People, cellData.Process)
			f.SetCellValue(sheetName, cell, value)

			color := model.FunctionColors[funcName]
			
			// Check if we need thick right border for "Protect" (column D)
			var borders []excelize.Border
			borders = append(borders, excelize.Border{Type: "left", Color: "000000", Style: 1})
			borders = append(borders, excelize.Border{Type: "top", Color: "000000", Style: 1})
			borders = append(borders, excelize.Border{Type: "bottom", Color: "000000", Style: 1})
			
			if funcName == "Protect" {
				borders = append(borders, excelize.Border{Type: "right", Color: "000000", Style: 5}) // 5 is thick
			} else {
				borders = append(borders, excelize.Border{Type: "right", Color: "000000", Style: 1})
			}

			style, _ := f.NewStyle(&excelize.Style{
				Alignment: &excelize.Alignment{Horizontal: "left", Vertical: "top", WrapText: true},
				Fill:      excelize.Fill{Type: "pattern", Color: []string{color}, Pattern: 1},
				Border:    borders,
			})
			f.SetCellStyle(sheetName, cell, cell, style)
		}
	}

	// 4. Thick border for Boom Boundary
	// Protect is column D (4). We apply it to rows 1 and 2 specifically.
	leftBoomStyleThick, _ := f.NewStyle(&excelize.Style{
		Fill:      excelize.Fill{Type: "pattern", Color: []string{"4F81BD"}, Pattern: 1},
		Font:      boomStyleBase.Font,
		Alignment: boomStyleBase.Alignment,
		Border: []excelize.Border{
			{Type: "left", Color: "000000", Style: 1},
			{Type: "top", Color: "000000", Style: 1},
			{Type: "bottom", Color: "000000", Style: 1},
			{Type: "right", Color: "000000", Style: 5},
		},
	})
	f.SetCellStyle(sheetName, "C1", "D1", leftBoomStyleThick)

	// Update Protect function header (D2)
	protectColor := model.FunctionColors["Protect"]
	protectHeaderStyle, _ := f.NewStyle(&excelize.Style{
		Font:      &excelize.Font{Bold: true, Size: 12},
		Alignment: &excelize.Alignment{Horizontal: "center", Vertical: "center", WrapText: true},
		Fill:      excelize.Fill{Type: "pattern", Color: []string{protectColor}, Pattern: 1},
		Border: []excelize.Border{
			{Type: "left", Color: "000000", Style: 1},
			{Type: "top", Color: "000000", Style: 1},
			{Type: "bottom", Color: "000000", Style: 1},
			{Type: "right", Color: "000000", Style: 5},
		},
	})
	f.SetCellStyle(sheetName, "D2", "D2", protectHeaderStyle)

	// 5. Add Dependency Legend
	legendRow := len(model.AssetClasses) + 4
	legendCell := fmt.Sprintf("A%d", legendRow)
	lastCol, _ := excelize.ColumnNumberToName(len(model.Functions) + 1)
	lastCell := fmt.Sprintf("%s%d", lastCol, legendRow)
	
	f.MergeCell(sheetName, legendCell, lastCell)
	f.SetCellValue(sheetName, legendCell, "Dependency emphasis: Govern is process-led | Identify/Protect lean Technology | Detect/Respond/Recover lean People")
	
	legendStyle, _ := f.NewStyle(&excelize.Style{
		Font:      &excelize.Font{Italic: true},
		Alignment: &excelize.Alignment{Horizontal: "center", Vertical: "center"},
	})
	f.SetCellStyle(sheetName, legendCell, legendCell, legendStyle)

	// Auto-adjust column widths (approximated)
	f.SetColWidth(sheetName, "A", "A", 15)
	f.SetColWidth(sheetName, "B", "G", 28)

	return f.SaveAs(path)
}
